onnxruntime/core/providers/cpu/llm/attention.cc

```diff
@@ -0,0 +1,749 @@
+  parameters.qk_matmul_output_mode = qk_matmul_output_mode_;                  // output mode for Q*K matmul
+  parameters.mask_type = attn_mask == nullptr ? AttentionMaskType::MASK_NONE  // mask type
+                                              : (attn_mask->IsDataType<bool>() ? AttentionMaskType::MASK_BOOL : AttentionMaskType::MASK_ADD);
+  parameters.batch_size = onnxruntime::narrow<int>(Q->Shape()[0]);  // Q.shape[0], K.shape[0], V.shape[0] (4D)
+
+  ORT_ENFORCE(parameters.batch_size > 0, "Batch size must be greater than 0");
+  ORT_ENFORCE(parameters.mask_type == AttentionMaskType::MASK_NONE ||
+                  parameters.mask_type == AttentionMaskType::MASK_BOOL ||
+                  attn_mask->IsDataType<float>(),
+              "Attention mask must be of type bool or float if provided.");
+
+  std::vector<int64_t> y_shape;
+  if (q_dims == 4) {
+    // 4D
+    parameters.kv_num_heads = kv_num_heads_ > 0 ? kv_num_heads_ : onnxruntime::narrow<int>(K->Shape()[1]);  // K.shape[1] or V.shape[1] (4D)
+    parameters.q_num_heads = q_num_heads_ > 0 ? q_num_heads_ : onnxruntime::narrow<int>(Q->Shape()[1]);     // Q.shape[1] (4D)
+
+    ORT_ENFORCE(parameters.kv_num_heads == onnxruntime::narrow<int>(K->Shape()[1]), "kv_num_heads different from K.shape[1]");
+    ORT_ENFORCE(parameters.kv_num_heads == onnxruntime::narrow<int>(V->Shape()[1]), "kv_num_heads different from V.shape[1]");
+    ORT_ENFORCE(parameters.q_num_heads == onnxruntime::narrow<int>(Q->Shape()[1]), "q_num_heads different from Q.shape[1]");
+    ORT_ENFORCE(Q->Shape()[3] == K->Shape()[3], "inconsistent head_size");
+    ORT_ENFORCE(K->Shape()[2] == V->Shape()[2], "inconsistent kv_sequence_length");
+    ORT_ENFORCE(attn_mask == nullptr || attn_mask->Shape()[attn_mask->Shape().NumDimensions() - 2] == Q->Shape()[2], "inconsistent q_sequence_length");
+    ORT_ENFORCE(past_key == nullptr || K->Shape()[1] == past_key->Shape()[1], "inconsistent kv_num_heads");
+    ORT_ENFORCE(past_value == nullptr || K->Shape()[1] == past_value->Shape()[1], "inconsistent kv_num_heads");
+    ORT_ENFORCE(past_key == nullptr || K->Shape()[3] == past_key->Shape()[3], "inconsistent head_size");
+    ORT_ENFORCE(past_value == nullptr || V->Shape()[3] == past_value->Shape()[3], "inconsistent head_size");
+
+    // From shapes
+    parameters.transpose_output = false;                                      // whether to transpose the output from BxNxSxH to BxSxNxH
+    parameters.q_sequence_length = onnxruntime::narrow<int>(Q->Shape()[2]);   // Q.shape[2] (4D)
+    parameters.head_size = onnxruntime::narrow<int>(Q->Shape()[3]);           // Q.shape[3] (4D)
+    parameters.kv_sequence_length = onnxruntime::narrow<int>(K->Shape()[2]);  // K.shape[2] or V.shape[2] (4D)
+    parameters.v_head_size = onnxruntime::narrow<int>(V->Shape()[3]);         // V.shape[3] (4D)
+    parameters.past_sequence_length = past_key == nullptr                     // past_key.shape[2] or past_value.shape[2] (4D) or given by the mask
```

Comment:
When this is set from past_key, should we check that past_value.shape[2] == past_key.shape[2]?

Reply:
done